### PR TITLE
Idcom 2172

### PR DESCRIPTION
**Pull request by ronald-hood-jr**
Draft PR: Contains wallet standard interface for Cryptid, Wallet Dapp with Basic Propose and Execute Transactions, and React Wallet Invoker that Generates the Iframe. I will create a README that explains the contents better and add comments throughout the repo.

### CODE REVIEW

File: wallet-standard/cryptid-wallet/package.json [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        "tsc": "tsc --build --verbose tsconfig.all.json",
+        "package": "shx echo '{ \"type\": \"commonjs\" }' > lib/cjs/package.json",
+        "build": "npm run clean && npm run tsc && npm run package"
+    },
+    "dependencies": {
+        "@solana/wallet-standard-features": "^1.0.0-rc.7",
+        "@solana/web3.js": "^1.58.0",
+        "@wallet-standard/base": "^1.0.0-rc.5",
+        "@wallet-standard/features": "^1.0.0-rc.5",
+        "@identity.com/cryptid": "0.2.0",
```
Comment (rado0x54) [inline wallet-standard/cryptid-wallet/package.json:32]: If do a `postMessage` conversation only, we would not expect a cryptid dependency here, correct? OR do we need any library logic on this site of the adapter?

File: wallet-standard/dapp/package.json [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  "private": false,
+  "scripts": {
+    "dev": "next dev",
+    "build": "next build",
+    "start": "next start",
+    "lint": "next lint"
+  },
+  "dependencies": {
+    "@heroicons/react": "^1.0.5",
+    "@identity.com/cryptid-idl": "^0.3.0-alpha.1",
```
Comment (rado0x54) [inline wallet-standard/dapp/package.json:14]: I would have expected a cryptid dependency here. (e.g. in order to get the CryptidAccount information)

Comment (rado0x54) [reply wallet-standard/dapp/package.json:14]: And I think the easiest integration of all Cryptid dependencies here would be to place it under the same workspace.

**Comment by JulianSpring** [comment]
Could you update the PR title to something more descriptive please :)

File: wallet-standard/cryptid-wallet/src/account.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+export class CryptidWalletWalletAccount implements WalletAccount {
+    readonly #address: WalletAccount['address'];
+    readonly #publicKey: WalletAccount['publicKey'];
+    readonly #chains: WalletAccount['chains'];
+    readonly #features: WalletAccount['features'];
+    readonly #label: WalletAccount['label'];
+    readonly #icon: WalletAccount['icon'];
+    _address: PublicKey;
```
Comment (JulianSpring) [inline wallet-standard/cryptid-wallet/src/account.ts:29]: Is this meant to be private? 
What is the difference between #address and _address?

File: wallet-standard/cryptid-wallet/src/account.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    didPrefix: string;
+    didAddress: string;
+    alias: string;
+    signer: Signer;
+    connection: Connection;
+    isSelected: (account: CryptidWalletWalletAccount) => boolean;
+    parent?: CryptidWalletWalletAccount;
+    _document: DIDDocument;
```
Comment (JulianSpring) [inline wallet-standard/cryptid-wallet/src/account.ts:37]: Is this meant to be private?

File: wallet-standard/cryptid-wallet/src/account.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        this.#chains = chains;
+        this.#features = features;
+        this.#label = label;
+        this.#icon = icon;
+    }
+    async init() {
+        this._address = await this.cryptid.address();
+        await this.updateDocument();
+        // console.log(`Getting address: ${this._address}`)
+        // console.log(`Getting document: ${JSON.stringify(this.document)}`)
```
Comment (JulianSpring) [inline wallet-standard/cryptid-wallet/src/account.ts:106]: Let's remove

File: wallet-standard/cryptid-wallet/src/wallet.ts [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+                version: '1.0.0',
+                signMessage: this.#signMessage,
+            },
+            'cryptidWallet:': {
+                cryptidWallet: this.#cryptidWallet,
+            },
+        };
+    }
+
+    get accounts() {
```
Comment (JulianSpring) [inline wallet-standard/cryptid-wallet/src/wallet.ts:102]: Return type

File: wallet-standard/cryptid-wallet/src/account.ts [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        await this.updateDocument();
+        return signature;
+    };
+
+    updateDocument = async () => {
+        this._document = await this.cryptid.document();
+        return this._document;
+    };
+
+    get address() {
```
Comment (JulianSpring) [inline wallet-standard/cryptid-wallet/src/account.ts:51]: Our functions should all have return types

File: wallet-standard/cryptid-wallet/src/wallet.ts [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+        } else if (inputs.length > 1) {
+            for (const input of inputs) {
+                outputs.push(...(await this.#signAndSendTransaction(input)));
+            }
+        }
+
+        return outputs;
+    };
+
+    #signTransaction: SolanaSignTransactionMethod = async (...inputs) => {
```
Comment (JulianSpring) [inline wallet-standard/cryptid-wallet/src/wallet.ts:209]: Can we get a type for `...inputs` ?

File: wallet-standard/cryptid-wallet/src/wallet.ts [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+            for (const input of inputs) {
+                if (input.account !== this.#account) throw new Error('invalid account');
+                if (input.chain) {
+                    if (!isSolanaChain(input.chain)) throw new Error('invalid chain');
+                    if (chain) {
+                        if (input.chain !== chain) throw new Error('conflicting chain');
+                    } else {
+                        chain = input.chain;
+                    }
+                }
```
Comment (JulianSpring) [inline wallet-standard/cryptid-wallet/src/wallet.ts:236]: We can use guard clauses and follow a 'happy path', generally nesting lots of if logic is hard to read :) 
 ( double check that i wrote this out correctly) 
```suggestion
                 if (input.account !== this.#account) throw new Error('invalid account'); 
                 
                 if (!input.chain) continue;
                 
                 if (!isSolanaChain(input.chain)) throw new Error('invalid chain');
                 
                 if (chain && input.chain !== chain) throw new Error('conflicting chain');
                 
                 chain = input.chain; 
```

File: wallet-standard/dapp/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+ENV NODE_ENV production
+
+COPY package*.json ./
+
+RUN npm ci 
+
+COPY . /opt/app
+
+RUN npm install --dev && npm run build
```
Comment (JulianSpring) [inline wallet-standard/dapp/Dockerfile:13]: Is `--dev` necessary?

**Review by JulianSpring — COMMENTED** [review]
Will review a bit more later. Please let me know if you wanna chat through anything :)

File: wallet-standard/cryptid-wallet/src/wallet.ts [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+    };
+
+    #signAndSendTransaction: SolanaSignAndSendTransactionMethod = async (...inputs) => {
+        if (!this.#account) throw new Error('not connected');
+
+        const outputs: SolanaSignAndSendTransactionOutput[] = [];
+
+        if (inputs.length === 1) {
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            const { transaction, account, chain, options } = inputs[0]!;
```
Comment (JulianSpring) [inline wallet-standard/cryptid-wallet/src/wallet.ts:184]: I don't think the non null assertion is necessary? We should never use non null in non tests as a general rule.

It looks like on `accounts` is the only var used here?

**Pull request closed without merging**